src/System.CommandLine/Binding/ArgumentConversionResult.cs

```diff
@@ -1,73 +1,94 @@
             Result = ArgumentConversionResultType.NoArgument;
         }
 
         internal ArgumentConversionResult(
-            Argument argument,
+            ArgumentResult argumentResult,
             Type expectedType,
-            string value,
-            LocalizationResources localizationResources) :
-            this(argument, FormatErrorMessage(argument, expectedType, value, localizationResources), ArgumentConversionResultType.FailedType)
+            string value) :
+            this(argumentResult, FormatErrorMessage(argumentResult, expectedType, value), ArgumentConversionResultType.FailedType)
         {
         }
 
-        internal static ArgumentConversionResult Failure(Argument argument, string error, ArgumentConversionResultType reason) => new(argument, error, reason);
+        internal static ArgumentConversionResult Failure(ArgumentResult argumentResult, string error, ArgumentConversionResultType reason)
+            => new(argumentResult, error, reason);
 
-        public static ArgumentConversionResult Success(Argument argument, object? value) => new(argument, value);
+        public static ArgumentConversionResult Success(ArgumentResult argumentResult, object? value)
+            => new(argumentResult, value);
 
-        internal static ArgumentConversionResult None(Argument argument) => new(argument);
+        internal static ArgumentConversionResult None(ArgumentResult argumentResult)
+            => new(argumentResult);
 
         private static string FormatErrorMessage(
-            Argument argument,
+            ArgumentResult argumentResult,
             Type expectedType,
-            string value,
-            LocalizationResources localizationResources)
+            string value)
         {
-            if (argument.FirstParent?.Symbol is IdentifierSymbol identifierSymbol &&
-                argument.FirstParent.Next is null)
+            if (argumentResult.Parent is CommandResult commandResult)
             {
-                var alias = identifierSymbol.GetLongestAlias(removePrefix: false);
+                string alias = commandResult.Command.GetLongestAlias(removePrefix: false);
+                CompletionItem[] completionItems = argumentResult.Argument.GetCompletions(CompletionContext.Empty).ToArray();
 
-                switch (identifierSymbol)
+                if (completionItems.Length > 0)
                 {
-                    case Command _:
-                        return localizationResources.ArgumentConversionCannotParseForCommand(value, alias, expectedType);
-                    case Option _:
-                        return localizationResources.ArgumentConversionCannotParseForOption(value, alias, expectedType);
+                    return argumentResult.LocalizationResources.ArgumentConversionCannotParseForCommand(
+                        value, alias, expectedType, completionItems.Select(ci => ci.Label));
```

Comment:
previously it was done by `ParseResultVisitor` (the error message was mutated, which did not play well with the new way of reporting erros) and was not using localization for the extra information received from completions. I moved entire logic here but this required having `ArgumentResult` and the information it provides

https://github.com/dotnet/command-line-api/blob/edd256b2fe63b57a79b39f10d0c63c81d43218ce/src/System.CommandLine/Parsing/ParseResultVisitor.cs#L522-L533

Reply:
Very nice.